server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1090,7 +1090,11 @@ public Searcher acquireSearcher(String source, SearcherScope scope, Function<Sea
         }
     }
 
-    protected abstract ReferenceManager<ElasticsearchDirectoryReader> getReferenceManager(SearcherScope scope);
+    protected abstract ReferenceManager<DirectoryReader> getReferenceManager(SearcherScope scope);
```

Comment:
All of these changes are needed because otherwise it is not possible to wrap readers since we are locked into `ElasticsearchDirectoryReader` which is final class.